Added ScanBy

leandromoh opened the pull request.

leandromoh commented:
@atifaziz tests were sucefull. The following error happened on linux build:

> $ export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
> $ export DOTNET_CLI_TELEMETRY_OPTOUT=1
> 6.11s0.47sE: Unable to locate package dotnet-sdk-2.1
> E: Couldn't find any package by glob 'dotnet-sdk-2.1'
> E: Couldn't find any package by regex 'dotnet-sdk-2.1'
> The command "sudo apt-get install -qq dotnet-sdk-2.1=2.1.300*" failed and exited with 100 during

atifaziz commented:
> The error on linux looks is about the build itself

@leandromoh Seems like it was an intermittent issue with OS X builds. Triggering a re-build [fixed the situation](https://travis-ci.org/morelinq/MoreLINQ/builds/522872135).

atifaziz commented on MoreLinq/ScanBy.cs:
```diff
@@ -0,0 +1,134 @@
+                var valueOfNullKey = (Value: default(TState), HasValue: false);
+                var state = default(TState);
+
+                bool TryGetValue(TKey key, out TState value)
```
For wording consistency & clarity, rename to `TryGetValue` to `TryGetState` and `value` to `state`.

leandromoh replied:
Method name was renamed but we cannot rename `value `to `state `because already exists a variable with this name in the outer scope (error CS0136).

atifaziz commented on MoreLinq/ScanBy.cs:
```diff
@@ -0,0 +1,134 @@
+                var stateByKey = new Dictionary<TKey, TState>(comparer);
+                var havePrevKey = false;
+                var prevKey = default(TKey);
+                var valueOfNullKey = (Value: default(TState), HasValue: false);
```
Better to use tuples or not but not a mix of both; that is, `valueOfNullKey` is a tuple but `havePrevKey` and `prevKey` are maintained in separate variables. So either break out the `valueOfNullKey` tuple like this:

```c#
var haveNullKeyState = false;
var nullKeyState = default(TState);
```

or use tuples defined like this:

```c#
var prevKey = (HasValue: false, Value: default(TKey));
var nullKeyState = (HasValue: false, Value: default(TState));
```

atifaziz commented on MoreLinq/ScanBy.cs:
```diff
@@ -0,0 +1,134 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
```
This is not used/needed.

atifaziz commented on MoreLinq/ScanBy.cs:
```diff
@@ -0,0 +1,134 @@
+        /// <summary>
+        /// Scans the source sequence and performs accumulations of state for each group of keys.
+        /// The resulting sequence's elements are generated by projecting each element of the
+        /// source sequece, its key and the current state for that key.
```
Typo: sequece

atifaziz commented on MoreLinq/ScanBy.cs:
```diff
@@ -0,0 +1,134 @@
+        /// <summary>
+        /// Scans the source sequence and performs accumulations of state for each group of keys.
+        /// The resulting sequence's elements are generated by projecting each element of the
+        /// source sequece, its key and the current state for that key.
```
Typo: sequece

atifaziz commented on README.md:
```diff
@@ -493,6 +493,15 @@ Peforms a scan (inclusive prefix sum) on a sequence of elements.
+
+Scans the source sequence and performs accumulations of state for each
+group of keys. The resulting sequence's elements are generated by
+projecting each element of the source sequece, its key and the current
```
Typo: sequece

leandromoh commented:
> @leandromoh Seems like it was an intermittent issue with OS X builds. Triggering a re-build fixed the situation.

How could I re-trigger the build? By the way, ajusts done.

atifaziz commented on MoreLinq/ScanBy.cs:
```diff
@@ -0,0 +1,131 @@
+                comparer = comparer ?? EqualityComparer<TKey>.Default;
+
+                var stateByKey = new Dictionary<TKey, TState>(comparer);
+                var prevKey = (Value: default(TKey), HasValue: false);
```
Reverse tuple elements in order of importance such that `HasValue` comes before `Value`.

leandromoh replied:
adjust done

atifaziz commented:
> How could I re-trigger the build?

You can do this by generating a blank commit, using the [`--allow-empty` switch of the `git commit` command](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---allow-empty). I don't need to do that because I have permissions to trigger and manage builds.

leandromoh commented:
@atifaziz adjusts done. Thanks for the invite to join you as MoreLinq's maintainer by the way.

leandromoh commented:
@atifaziz adjusts done. Thanks for the invite to join you as MoreLinq's collaborator. However, I did not notice nothing different. What changed?

atifaziz commented on MoreLinq.Test/ScanByTest.cs:
```diff
@@ -0,0 +1,142 @@
+            var result = source.ScanBy(
+                item =>
+                {
+                    Assert.That(item, Is.EqualTo(source[sourceIndex]));
```
This assertion within lambdas and others alike seems a little excessive because I am not sure they are helping to test anything beyond what you get from comparing against the expected result. I'd therefore keep it simple, as follows:

```c#
[Test]
public void ScanBy()
{
    var source = new[]
    {
        "ana",
        "beatriz",
        "carla",
        "bob",
        "davi",
        "adriano",
        "angelo",
        "carlos"
    };

    var result =
        from x in
            source.ScanBy(
                item => item.First(),
                key => (Element: default(string), State: -1),
                (item, key, state) => (Element: item, State: state.State + 1))
        select (x.Value.Element, x.Key, x.Value.State);

    result.AssertSequenceEqual(
        ("ana",     'a', 0),
        ("beatriz", 'b', 0),
        ("carla",   'c', 0),
        ("bob",     'b', 1),
        ("davi",    'd', 0),
        ("adriano", 'a', 1),
        ("angelo",  'a', 2),
        ("carlos",  'c', 1));
}
```

atifaziz commented:
> Thanks for the invite to join you as MoreLinq's collaborator. However, I did not notice nothing different. What changed?

@leandromoh Sorry, my mistake. I have hopefully fixed this. You can now help triage/manage issues, can be assigned issues, and more importantly, requested a review. Being the second highest contributor to the project, I feel you're very familiar with the code base and the project by now so I value your comments with regards to contributions from others. Going through a PR is still the best way to contribute code, however.

atifaziz commented on MoreLinq/ScanBy.cs:
```diff
@@ -0,0 +1,130 @@
+            this IEnumerable<TSource> source,
+            Func<TSource, TKey> keySelector,
+            Func<TKey, TState> seedSelector,
+            Func<TSource, TKey, TState, TState> accumulator,
```
For consistency with `Scan`, the type arguments of `accumulator` should be reversed to be `Func<TState, TKey, TSource, TState>`.

leandromoh replied:
adjusted in [#d6f5755](https://github.com/morelinq/MoreLINQ/pull/572/commits/d6f5755ad1aea7b3d4fb9d702c58e55a171c094d)

leandromoh commented:
@atifaziz adjust done about accumulator. I simplified that test that had assert on lambdas, but notice that it was very fragile since we didnt use the `key` passed to `seedSelector` netheir `accumulator`, and therefore could not assert properly. So I adjusted to use `key` as part of `state` on test, and now every single parameter is being validated without asserts inside lambdas.

atifaziz commented on MoreLinq.Test/ScanByTest.cs:
```diff
@@ -0,0 +1,119 @@
+            var result =
+                source.ScanBy(
+                          item => item.First(),
+                          key => (Element: default(string), Key: key, State: (int)key - 1),
```
The cast in `(int)key - 1` is redundant.

atifaziz commented on MoreLinq.Test/ScanByTest.cs:
```diff
@@ -0,0 +1,120 @@
+                from x in
+                    source.ScanBy(
+                        item => item.First(),
+                        key => (Element: default(string), Key: key, State: (int)key - 1),
```
The cast in `(int)key - 1` is redundant.

atifaziz commented on MoreLinq.Test/ScanByTest.cs:
```diff
@@ -0,0 +1,120 @@
+                    source.ScanBy(
+                        item => item.First(),
+                        key => (Element: default(string), Key: key, State: (int)key - 1),
+                        (state, key, item) => (Element: item, Key: char.ToUpper(key), State: state.State + 1))
```
- Replace `ToUpper` with the more deterministic `ToUpperInvariant`.
- The tuple element names here are redundant. I think it reads equally clear if we remove them, as in `(item, char.ToUpper(key), state.State + 1)`.

atifaziz commented on MoreLinq.Test/ScanByTest.cs:
```diff
@@ -0,0 +1,120 @@
+                        item => item.First(),
+                        key => (Element: default(string), Key: key, State: (int)key - 1),
+                        (state, key, item) => (Element: item, Key: char.ToUpper(key), State: state.State + 1))
+                select (x.Key, (x.Value.Element, x.Value.Key, x.Value.State - x.Key));
```
I'm not sure what you intended with the `x.Value.State - x.Key` expression. Removing the subtraction of `x.Key` doesn't change anything. The test still passes so it tells me that it's not serving any purpose. In fact, the whole of the last projection can be removed and so the test code just becomes as simple as:

```c#
            var result =
                source.ScanBy(
                    item => item.First(),
                    key => (Element: default(string), Key: key, State: (int)key - 1),
                    (state, key, item) => (item, char.ToUpper(key), state.State + 1));

            result.AssertSequenceEqual(
               KeyValuePair.Create('a', ("ana",     'A', 0)),
               KeyValuePair.Create('b', ("beatriz", 'B', 0)),
               KeyValuePair.Create('c', ("carla",   'C', 0)),
               KeyValuePair.Create('b', ("bob",     'B', 1)),
               KeyValuePair.Create('d', ("davi",    'D', 0)),
               KeyValuePair.Create('a', ("adriano", 'A', 1)),
               KeyValuePair.Create('a', ("angelo",  'A', 2)),
               KeyValuePair.Create('c', ("carlos",  'C', 1)));
```

atifaziz replied:
> @atifaziz without the subtraction on last projection 0 would become 97 ('a'), 1 would be 98 ('b') e so on. Remember that seedSelector now returns the int value for each key as seed.

Sure but my point is that the test passes regardless so it can't be serving any intended purpose.

leandromoh replied:
[#f70ee87](https://github.com/morelinq/MoreLINQ/pull/572/commits/f70ee87f2f6ade813668aee8835d3dfa67066455) fix this and the other resquests.

leandromoh replied:
[#f70ee87](https://github.com/morelinq/MoreLINQ/pull/572/commits/f70ee87f2f6ade813668aee8835d3dfa67066455) fixed this and the other resquests (removed cast; ToUpperInvariant; removed last projection, unnecessary tuple element names). sorry for the sigle commit, I did not have much free time.

leandromoh commented:
@atifaziz without the subtraction on last projection 0 would become 64, 1 would be 65 e so on. Remember that seedSelector now returns the int value for each key as seed.

atifaziz commented:
📝 This PR is for #563.
